File: src/www/__test__/blog/Post_20220715_HelloBlog_Context.js
import { queryAllByTestId, getByTestId } from "@testing-library/react";
import { clickLink } from "../main/actions";
import { waitForLoading } from "../loading/actions";
import { mainView } from "../main";

export class Post_20220715_HelloBlog_Context {
  async beforeTest() {}

  async goToTheBlogSection() {
    clickLink(mainView, "Blog");
    await waitForLoading();
  }

  async youShouldSeeAListOfPosts() {
    const posts = queryAllByTestId(mainView, "post-list--item");
    expect(posts).not.toHaveLength(0);
  }

  async theLastPostTitleShouldBeSThisPost(expected) {
    // expected = "Hello Blog"

    const posts = queryAllByTestId(mainView, "post-list--item");
    const last = posts.at(-1);

    expect(last).toHaveTextContent(expected);
  }

  async goToTheSPost(postTitle) {
    clickLink(mainView, postTitle);
    await waitForLoading();
  }

  async youShouldSeeTheSPost(the) {
    // the = "Hello Blog"

    const title = getByTestId(mainView, "post-title");
    expect(title).toHaveTextContent(the);
  }

  async thePostShouldContainSWhichIsHere(contain) {
    const body = getByTestId(mainView, "post-body");
    expect(body).toHaveTextContent(contain);
  }

  async afterTest() {}
}

File: src/test/java/com/drpicox/game/blog/Post_20220715_HelloBlog_Context.java
package com.drpicox.game.blog;

import com.drpicox.game.blog.api.ListPostsDTO;
import com.drpicox.game.blog.api.ListPostsEntryDTO;
import com.drpicox.game.blog.api.PostDTO;
import org.springframework.stereotype.Component;
import static com.google.common.truth.Truth.assertThat;
import static com.google.common.truth.Truth8.assertThat;
import com.drpicox.game.util.FrontendSimulator;

import java.util.Optional;

@Component
public class Post_20220715_HelloBlog_Context {

    private final FrontendSimulator frontendSimulator;

    private ListPostsDTO postsListDTO;
    private PostDTO postDTO;

    Post_20220715_HelloBlog_Context(FrontendSimulator frontendSimulator) {
        this.frontendSimulator = frontendSimulator;
    }

    public void beforeTest() {
    }

    public void goToTheBlogSection() {
        postsListDTO = frontendSimulator.get("/api/v1/posts", ListPostsDTO.class);
    }

    public void youShouldSeeAListOfPosts() {
        assertThat(postsListDTO.getPosts()).isNotEmpty();
    }

    public void theLastPostTitleShouldBeSThisPost(String expected) {
        // expected = "Hello Blog"
        var query = findPostEntry(expected);
        assertThat(query).isPresent();
    }

    public void goToTheSPost(String the) {
        // the = "Hello Blog"
        var entry = findPostEntry(the);
        var id = entry.get().getId();
        postDTO = frontendSimulator.get("/api/v1/posts/" + id, PostDTO.class);
    }

    public void youShouldSeeTheSPost(String the) {
        assertThat(postDTO.getTitle()).isEqualTo(the);
    }

    public void thePostShouldContainSWhichIsHere(String contain) {
        assertThat(postDTO.getBody()).contains(contain);
    }

    private Optional<ListPostsEntryDTO> findPostEntry(String expectedTitle) {
        return postsListDTO.getPosts().stream().filter(p -> p.getTitle().equals(expectedTitle)).findAny();
    }

    public void afterTest() {
    }
}
